src/easy_mark_editor.rs

```diff
@@ -0,0 +1,274 @@
+use eframe::egui::{text_edit::CCursorRange, *};
+use eframe::{egui, epi};
+use rfd::{FileDialog, MessageDialog};
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::PathBuf;
+
+pub struct TextEditor {
+    code: String,
+    show_rendered: bool,
+    file_path: Option<PathBuf>,
+    should_exit: bool,
+    is_exiting: bool,
+}
+
+impl PartialEq for TextEditor {
+    fn eq(&self, other: &Self) -> bool {
+        (&self.code, self.show_rendered) == (&other.code, other.show_rendered)
+    }
+}
+
+impl Default for TextEditor {
+    fn default() -> Self {
+        Self {
+            code: String::new(),
+            show_rendered: true,
+            file_path: None,
+            should_exit: false,
+            is_exiting: false,
+        }
+    }
+}
+
+impl epi::App for TextEditor {
+    fn on_exit_event(&mut self) -> bool {
+        self.is_exiting = true;
+        self.should_exit
+    }
+
+    fn name(&self) -> &str {
+        "Rust text editor"
+    }
+
+    fn update(&mut self, ctx: &eframe::egui::Context, frame: &epi::Frame) {
+        ctx.set_visuals(egui::Visuals::dark());
+        egui::SidePanel::right("side_panel").show(ctx, |ui| {
+            if ui.button(format!("{:^17}", "Quit")).clicked() {
+                eprintln!("Quitting via the 'Quit' button");
+                frame.quit();
+            }
+
+            if ui.button(format!("{:^13}", "Open file")).clicked() && matches!(self.open(), Err(_))
+            {
+                MessageDialog::new()
+                    .set_title("File is opening")
+                    .set_description("Failed to open the file")
+                    .set_buttons(rfd::MessageButtons::Ok)
+                    .show();
+            }
+
+            if ui.button(format!("{:^15}", "Save file")).clicked()
+                && matches!(self.save(false), Err(_))
+            {
+                MessageDialog::new()
+                    .set_title("File is saving")
+                    .set_description("Failed to save the file")
+                    .set_buttons(rfd::MessageButtons::Ok)
+                    .show();
+            }
+
+            if ui.button(format!("{:^13}", "Save file as")).clicked()
+                && matches!(self.save(true), Err(_))
+            {
+                MessageDialog::new()
+                    .set_title("Saving file as")
+                    .set_description("Failed to save the file")
+                    .set_buttons(rfd::MessageButtons::Ok)
+                    .show();
+            }
+
+            if ui.input_mut().consume_key(egui::Modifiers::COMMAND, Key::Q) {
+                eprintln!("Quitting via ctrl-q");
+                frame.quit();
+            }
+
+            if self.is_exiting {
+                self.should_exit = self.quit();
```

Comment:
jakieś to się wydaje przekombinowane i nie jest wytłumaczone czemu tak jest zrobione